pandas/core/strings.py

```diff
@@ -418,38 +418,129 @@ def _get_single_group_name(rx):
         return None
 
 
-def str_extract(arr, pat, flags=0):
+def _str_extract_noexpand(arr, pat, flags=0):
     """
     Find groups in each string in the Series using passed regular
-    expression.
+    expression. This function is called from
+    str_extract(expand=False), and can return Series, DataFrame, or
+    Index.
+
+    """
+    from pandas.core.frame import DataFrame
+    from pandas.core.index import Index
+
+    regex = re.compile(pat, flags=flags)
+    # just to be safe, check this
+    if regex.groups == 0:
+        raise ValueError("This pattern contains no groups to capture.")
+    empty_row = [np.nan] * regex.groups
+
+    def f(x):
+        if not isinstance(x, compat.string_types):
+            return empty_row
+        m = regex.search(x)
+        if m:
+            return [np.nan if item is None else item for item in m.groups()]
+        else:
+            return empty_row
+
+    if regex.groups == 1:
+        result = np.array([f(val)[0] for val in arr], dtype=object)
+        name = _get_single_group_name(regex)
+    else:
+        if isinstance(arr, Index):
+            raise ValueError("only one regex group is supported with Index")
+        name = None
+        names = dict(zip(regex.groupindex.values(), regex.groupindex.keys()))
+        columns = [names.get(1 + i, i) for i in range(regex.groups)]
+        if arr.empty:
+            result = DataFrame(columns=columns, dtype=object)
+        else:
+            result = DataFrame([f(val) for val in arr], columns=columns,
+                               index=arr.index, dtype=object)
+    return result, name
+
+
+def _str_extract_frame(arr, pat, flags=0):
+    """
+    For each subject string in the Series, extract groups from the
+    first match of regular expression pat. This function is called from
+    str_extract(expand=True), and always returns a DataFrame.
+
+    """
+    from pandas import Series, DataFrame, Index
+
+    regex = re.compile(pat, flags=flags)
+
+    if regex.groups == 0:
+        raise ValueError("pattern contains no capture groups")
+
+    names = dict(zip(regex.groupindex.values(), regex.groupindex.keys()))
+    columns = [names.get(1 + i, i) for i in range(regex.groups)]
+    if len(arr) == 0:
+        return DataFrame(columns=columns, dtype=object)
+
+    empty_row = [np.nan] * regex.groups
+
+    def f(x):
```

Comment:
*might* worth it to separate out this ``f(x)`` to an external function factory to generate for ``_str_extract_noexpand`` and here to not repeat the function